cudaextender/src/ungapped_xdrop_kernels.cu

```diff
@@ -0,0 +1,459 @@
+                                                int32_t* d_done)
+{
+    constexpr int32_t num_warps = 4;  // TODO - move out?
+    constexpr int32_t nuc       = 8;  // TODO - remove hardcode - pass in
+    constexpr int32_t nuc2      = 64; // TODO - remove hardcode
+    const int32_t lane_id       = threadIdx.x % warpSize;
+    const int32_t warp_id       = (threadIdx.x - lane_id) / warpSize;
+    const float log_4           = log(4.0f);
+    __shared__ int32_t ref_loc[num_warps];
+    __shared__ int32_t query_loc[num_warps];
+    __shared__ int32_t total_score[num_warps];
+    __shared__ int32_t prev_score[num_warps];
+    __shared__ int32_t prev_max_score[num_warps];
+    __shared__ int32_t prev_max_pos[num_warps];
+    __shared__ bool edge_found[num_warps];
+    __shared__ bool xdrop_found[num_warps];
+    __shared__ bool new_max_found[num_warps];
+    __shared__ int32_t left_extent[num_warps];
+    __shared__ int32_t extent[num_warps];
+    __shared__ int32_t tile[num_warps];
+    __shared__ double entropy[num_warps]; // TODO - Can this be a float?
+    __shared__ int32_t sub_mat[nuc2];
+
+    if (threadIdx.x < nuc2)
+    {
+        sub_mat[threadIdx.x] = d_sub_mat[threadIdx.x];
+    }
+    __syncthreads();
+
+    for (int32_t hid0 = blockIdx.x * num_warps; hid0 < num_seed_pairs; hid0 += num_warps * gridDim.x)
+    {
+        short count[4]     = {0};
+        short count_del[4] = {0};
+        const int32_t hid  = hid0 + warp_id + start_index;
+        if (lane_id == 0)
+        {
+            if (hid < num_seed_pairs)
+            {
+                ref_loc[warp_id]   = d_seed_pairs[hid].target_position_in_read;
+                query_loc[warp_id] = d_seed_pairs[hid].query_position_in_read;
+            }
+            else
+            {
+                ref_loc[warp_id]   = d_seed_pairs[hid0].target_position_in_read;
+                query_loc[warp_id] = d_seed_pairs[hid0].query_position_in_read;
+            }
+            total_score[warp_id]    = 0;
+            tile[warp_id]           = 0;
+            xdrop_found[warp_id]    = false;
+            edge_found[warp_id]     = false;
+            new_max_found[warp_id]  = false;
+            entropy[warp_id]        = 1.0f;
+            prev_score[warp_id]     = 0;
+            prev_max_score[warp_id] = -1000;
+            prev_max_pos[warp_id]   = 0;
+            extent[warp_id]         = 0;
+        }
+        __syncwarp();
+        //////////////////////////////////////////////////////////////////
+        //Right extension
+        while (!xdrop_found[warp_id] && !edge_found[warp_id])
+        {
+            int32_t max_pos;
+            int32_t thread_score = 0;
+            int32_t max_thread_score;
+            const int32_t pos_offset = lane_id + tile[warp_id];
+            const int32_t ref_pos    = ref_loc[warp_id] + pos_offset;
+            const int32_t query_pos  = query_loc[warp_id] + pos_offset;
+            char r_chr;
+            char q_chr;
```

Comment:
It seems `r_chr` and `q_chr` could be used uninitialized in line 209.
Haven't checked the full algorithms yet, but maybe the `if` in line 113 should not conditionally initialize the variables, but instead be modified to abort the while-loop or something?

Reply:
Good catch! I think here I might have to introduce warp active masking or else the syncwarps will throw it off.

Reply:
Or I can use a canary (`CHAR_MIN/MAX`).

Reply:
@ahehn-nv so if I understand correctly, you are proposing the initialization of `r_chr` and `q_chr` before the `if` in 113?

Reply:
Not, quite, but let's have a discussion about the algorithmic ideas behind the kernel first and address this in a separate PR.
(I'll leave the conversation unresolved as a reminder.)

Reply:
Filed this issue to track: https://github.com/clara-parabricks/GenomeWorks/issues/579